### PR TITLE
Allow disabling the AssetTargetFallback dependency resolution via an environment variable

### LINKED ISSUE #11564: [Bug]: AssetTargetFallback now brings in new dependencies and is causing build failures - Add a means to revert to the 17.1 behavior
### NuGet Product Used

MSBuild.exe

### Product Version

Microsoft (R) Build Engine version 17.2.0-preview-22081-04+288ea72fc for .NET Framework

### Worked before?

Microsoft (R) Build Engine version 17.1.0+ae57d105c for .NET Framework

### Impact

I'm unable to use this version

### Repro Steps & Context

As of the latest version, it seems like a missing dependency group breaks restore.

When packing, I do see this warning:
```
WARNING: NU5128: Some target frameworks declared in the dependencies group of the nuspec and the lib/ref folder do not have exact matches in the other location. Consult the list of actions below:
- Add a dependency group for .NETStandard2.0 to the nuspec
```

However, previously the behavior of a missing dependency group matched that of an empty dependency group.

Repro materials: [Repro.zip](https://github.com/NuGet/Home/files/8005402/Repro.zip)

Repro steps:
1. msbuild /restore TestLib
2. nuget pack -OutputDirectory Packages TestPackageBroken.nuspec
3. msbuild /t:restore TestProjectBroken

A warning is emitted:
```
C:\Code\tmp\RestoreRepro\TestProjectBroken\TestProjectBroken.csproj : warning NU1701: Package 'Microsoft.AspNet.WebApi.Core 5.2.7' was restored using '.NETFramework,Version=v4.6.1, .NETFramework,Version=v4.6.2, .NETFramewo
rk,Version=v4.7, .NETFramework,Version=v4.7.1, .NETFramework,Version=v4.7.2, .NETFramework,Version=v4.8' instead of the project target framework '.NETStandard,Version=v2.0'. This package may not be fully compatible with your
 project.
```

It works if you do this however (the package has an empty dependency group):
1. msbuild /restore TestLib
2. nuget pack -OutputDirectory Packages TestPackageWorks.nuspec
3. msbuild /t:restore TestProjectWorks

### Verbose Logs

_No response_

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/11564

Regression? Last working version:

## Description

Some users have been affected by the recent bug fix to start resolving the dependencies via AssetTargetFallback. In particular, https://github.com/NuGet/NuGet.Client/pull/4372/files is the change. 

https://github.com/NuGet/Home/issues/11564 has testimonials about it, and as a response we will:
* Disable the behavior in 17.3/6.3/6.0.400. 
* Remove the fallback in 17.4/6.4/7.0.100. 

This should ease the transition for the more severely affected customers. 

Normally the most convenient way of disabling this would be a property in the project, but that change creates a bunch of public API changes that we'd need to keep. 
Current NuGet does a lot of heavy caching when running restore, to ensure that we don't do multiple assets selections for a package id/version/framework. In order to ensure correctness, we'd have to account for this extra variable as well, whether ATF dependency resolution is enabled or not. 
I'd argue this is risky and especially given that we'd want to revert this in 17.4, I think the cost would be too high to do it on a project level.  
For reference this is how the other change would look like: https://github.com/NuGet/NuGet.Client/compare/dev-nkolev92-ATFDependenciesBehaviorFallback?expand=1.

The env var name is NUGET_USE_LEGACY_ASSET_TARGET_FALLBACK_DEPENDENCY_RESOLUTION and only the `true` case ignored value will be considered.

 
## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled - TBD. Waiting on approval from @aortiz-msft on this.
  - **OR**
  - [ ] N/A

### CODE REVIEW

File: test/NuGet.Core.Tests/NuGet.Commands.Test/SourceRepositoryDependencyProviderTests.cs [hunk 1]
```diff
@@ -772,6 +774,141 @@ public async Task FindLibraryAsync_WhenASourceIsInaccessible_AndFailuresAreIgnor
...
+                    packageDependencyGroups,
+                    Enumerable.Empty<FrameworkSpecificGroup>()));
+
+            var source = new Mock<SourceRepository>();
+            source.Setup(s => s.GetResourceAsync<FindPackageByIdResource>())
+                .ReturnsAsync(findResource.Object);
+            source.SetupGet(s => s.PackageSource)
+                .Returns(new PackageSource("http://test/index.json"));
+
+            var libraryRange = new LibraryRange(
```
Comment (erdembayar) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/SourceRepositoryDependencyProviderTests.cs]: Is this `libraryRange` used somewhere?

File: test/NuGet.Core.Tests/NuGet.Commands.Test/SourceRepositoryDependencyProviderTests.cs [hunk 2]
```diff
@@ -772,6 +774,141 @@ public async Task FindLibraryAsync_WhenASourceIsInaccessible_AndFailuresAreIgnor
...
```
Comment (erdembayar) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/SourceRepositoryDependencyProviderTests.cs]: Same here. Is it used after assignment?

File: test/NuGet.Core.Tests/NuGet.Commands.Test/SourceRepositoryDependencyProviderTests.cs (hunk 1 above)
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/SourceRepositoryDependencyProviderTests.cs]: Nope, remnant of a previous test version.
nice catch.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/SourceRepositoryDependencyProviderTests.cs (hunk 2 above)
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/SourceRepositoryDependencyProviderTests.cs]: Removed.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs [hunk 3]
```diff
@@ -92,7 +93,27 @@ public class SourceRepositoryDependencyProvider : IRemoteDependencyProvider
...
+                ignoreFailedSources,
+                ignoreWarning,
+                fileCache,
+                isFallbackFolderSource,
+                environmentVariableReader: EnvironmentVariableWrapper.Instance)
+        {
+        }
+
+        internal SourceRepositoryDependencyProvider(
+        SourceRepository sourceRepository,
```
Comment (jeffkl) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs]: nit: indentation is off

**Review by jeffkl — COMMENTED** [review]
In my opinion, having to set an environment variable is a little rough.  Say my entire repository isn't working with ATF yet, I can't expect everyone who clones and builds it to set an environment variable.  It would be much better if it was an MSBuild property I could set instead but of course that's a lot more work for us 🤕

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs]: fixed.

**Comment by nkolev92** [comment]
> In my opinion, having to set an environment variable is a little rough. Say my entire repository isn't working with ATF yet, I can't expect everyone who clones and builds it to set an environment variable. It would be much better if it was an MSBuild property I could set instead but of course that's a lot more work for us 🤕

Yeah, it creates maintainability issues down the line if we did properties :(

Fortunately, I don't expect this to be used heavily.

**Review by jeffkl — APPROVED** [review]
An env var is better than nothing but I still think a property would be best

**Pull request merged**